src/mem/cache/prefetch/indirect_memory_v2.hh

```diff
@@ -0,0 +1,662 @@
+ *      address = base_address + (value << shift)
+ *
+ * Therefore, to calculate the prefetch address for array A, i.e. address, we
+ * need to calculate the base_address and shift. The IMP algorithm performs
+ * this calculation by solving the above equation using two pairs of
+ * (value, address). Thus, the problem now gets reduced to solving a system of
+ * linear equations with two variables. The solution to this linear system is
+ * the predicted pair (base_address, shift).
+ *
+ * Once the (base_address, shift) pair is known, then IMP can compute the
+ * predicted access to array A, i.e. address, whenever it observes a load to
+ * the indexing array B, i.e. value. Note that for this algorithm to work IMP
+ * must figure out the value of index, i.e. the value loaded from array B.
+ * Therefore, IMP must track access to array B and then uses these accesses,
+ * along with access to array A, to compute the indirect access parameters
+ * (base_address, shift). IMP uses a stream detector to track accesses to array
+ * B. For computing the indirect access parameters, IMP uses a structure called
+ * the IndirectPatternDetector (IPD). After IMP's IPD detects an indirect
+ * pattern, i.e. the once it is able to compute the indirect parameters, IMP
+ * stores the computed pair in its IndirectTable
+ *
+ * For more details about the IMP algorithm, please refer to the paper
+ * mentioned in the references.
+ *
+ * What does this implementation DOESN'T do:
+ * There are some differences between this implementation and the algorithm
+ * described in the paper. Namely, this implementation does not support
+ * indirect prefetching for multi-way and multi-level accesses as described in
+ * section 3.3.2 of the paper
+ */
+
+#ifndef __MEM_CACHE_IMP_HH__
+#define __MEM_CACHE_IMP_HH__
+
+#include <map>
+#include <optional>
+#include <utility>
+#include <vector>
+
+#include "mem/cache/prefetch/queued.hh"
+#include "mem/packet.hh"
+#include "mem/packet_access.hh"
+#include "params/IMPv2Prefetcher.hh"
+#include "sim/system.hh"
+
+namespace gem5
+{
+    namespace prefetch
+    {
+        namespace IMPv2Internals
+        {
+            template<typename T>
+            void saturating_increment(T& old_value, const T& upper_bound)
+            {
+                if (old_value < upper_bound)
+                    old_value += 1;
+            }
+
+            template<typename T>
+            void saturating_decrement(T& old_value)
+            {
+                if (old_value > 0)
+                    old_value -= 1;
+            }
+
+            /*
+             * The StreamDetector performs two functions:
+             *  1. It detects stream accesses to the indexing array, B
+             *  2. If it is confident about a stream, it returns the predicted
+             *     delta for that stream so that IMP may issue stream
+             *     prefetches using the delta
+             */
+            class StreamDetector
```

Comment:
Can you reuse the Stride prefetcher instead of implementing this class? IIRC, a stream prefetcher is a stride prefetcher where the degree and the distance are 1, right? Then to get the outcome of inStreamTable you would instead call calculatePrefetch.

Reply:
I have a strong opinion against duplicating code, specially given that I spent years deduplicating stuff in the cache directory, so I will double down on this request.

I understand that it does more stuff, but isn´t there a combination of parameters that ends up generating the addresses that you want? I am leaning towards thinking that the extra stuff will just slightly degrade performance, but won´t do something irreversible to the state, or end up generating different addresses.